File: backend/app/api/routes/notifications.py
from typing import List
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from app.api.deps import get_db, get_current_user
from app.models.user import User
from app.models.notification import Notification
from app.schemas.notification import NotificationResponse, UnreadCountResponse

router = APIRouter()

@router.get("/", response_model=List[NotificationResponse])
def get_notifications(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return db.query(Notification).filter(Notification.recipient_id == current_user.id).order_by(Notification.created_at.desc()).all()

@router.get("/unread-count", response_model=UnreadCountResponse)
def get_unread_count(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    count = db.query(Notification).filter(Notification.recipient_id == current_user.id, Notification.is_read == False).count()
    return UnreadCountResponse(count=count)

@router.patch("/{notification_id}/read", response_model=NotificationResponse)
def mark_as_read(notification_id: int, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    notification = db.query(Notification).filter(Notification.id == notification_id, Notification.recipient_id == current_user.id).first()
    if notification:
        notification.is_read = True
        db.commit()
        db.refresh(notification)
    return notification

@router.post("/read-all", response_model=dict)
def mark_all_as_read(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    db.query(Notification).filter(Notification.recipient_id == current_user.id, Notification.is_read == False).update({"is_read": True})
    db.commit()
    return {"status": "success"}

File: frontend/src/api/notifications.ts
import api from './client';

export interface Notification {
  id: number;
  notification_type: string;
  title: string;
  message: string;
  related_entity_ref?: string;
  is_read: boolean;
  created_at: string;
}

export const notificationsApi = {
  getNotifications: async () => {
    const response = await api.get('/notifications/');
    return response.data as Notification[];
  },
  getUnreadCount: async () => {
    const response = await api.get('/notifications/unread-count');
    return response.data;
  },
  markAsRead: async (id: number) => {
    const response = await api.patch(`/notifications/${id}/read`);
    return response.data;
  },
  markAllAsRead: async () => {
    const response = await api.post('/notifications/read-all');
    return response.data;
  }
};
